### DISCUSSION TITLE
Different Homing Switch Pull-off Distance for Each Axis?

### DISCUSSION THREAD

**Discussion by hollingerc** [grblHAL/core#619] [Q&A] [answered]
Is there a way to have different pull-off distances for each axis?  My Z-axis works differently to the X and Y axes and needs a much larger distance, about 20X more.

**Comment by terjeio** [comment]
It is not possible without changing the core, if you compile yourself then modify the distance when z is commanded to move [here](https://github.com/grblHAL/core/blob/1444c08997827a5653eea821f4739fdc8a689a78/machine_limits.c#L187-L190).

Adding separate settings for each axis is relatively easy to do from a coding standpoint by not so when considering sender compatibility and when loading updated firmware on a controller with a previous settings structure.
I'll think about changing the core to allow a plugin to add new settings for those who needs it allowing the current setting to remain as-is for backwards compatibility.

**Reply by hollingerc** [reply]
Thanks, but I'll leave modifying the core up to you.  I have a method that gets around this issue and is working for now.

If you do change the core to allow a plugin, I'd be interested.  How would I know if/when you have done this?

**Reply by terjeio** [reply]
I have now modified the core and also [added a plugin](https://github.com/grblHAL/Plugins_misc/tree/main?tab=readme-ov-file#homing-pulloff) for you to test.

**Reply by hollingerc** [reply]
Great!  For me, at least, this will be useful.  I'll try it in a week or so, I have other priorities to deal with first.

Thanks for adding the plugin.

**Reply by hollingerc** [reply]
I setup the $180-182 and $190-192 configurations, but that didn't work.  The $24 and $25 settings seemed to take precedence.  Is there a macro that has to be enabled for this to work?

**Reply by terjeio** [reply]
> ... and $190-192 configurations

The pullof settings are $290 - 292, you have to [enable the plugin](https://github.com/grblHAL/Plugins_misc/tree/main?tab=readme-ov-file#homing-pulloff) as described in the readme.

**Reply by hollingerc** [reply]
I was referring to the homing speed settings for each axis - $180-182 homing locate feed rate and $190-192 homing search seek rate.  These were visible in the previous version of firmware I had before the new plugin.  I was able to manually change them and grbl remembered them.  Is the functionality of these settings enabled with the HOMING_PULLOFF_ENABLE macro?  Or is there another macro?  I did some searching for info on these settings but found only their descriptions.

BTW, I'm a getting linker (not compiler) error with the new code "undefined reference to 'board_init'".  I can give you more information if you need it.

**Reply by terjeio** [reply]
$18x and $19x is added by the motors plugin (Trinamic). Which board are you compiling for now?

**Reply by hollingerc** [reply]
I have my own custom board using different pinouts from the ST Morpho configuration.  I'm using STM32CubeIDE and selected the 'Release F411 Nucleo64' build configuration.  In the boards folder I created a my_machine_map.h file by copying, then modifying the contents of st_morpho_map.h.

**Reply by terjeio** [reply]
Ok, then remove `#define HAS_BOARD_INIT` from the map file if you are not using Trinamic drivers.

**Reply by hollingerc** [reply]
I commented out the HAS_BOARD_INIT macro and the code compiled without errors.

I am using Trinamic stepper motor drivers, three TMC2209s. The macros TRINAMIC_ENABLE 2209 and TRINAMIC_R_SENSE 110 are uncommented in my_machine.h.  The HAS_BOARD_INIT  macro was enabled when I compiled the previous version of grblHAL (pre-homing pulloff plugin) and there were no issues with compiling.

I don't have time to re-flash the grbl MCU right now, perhaps later today or tomorrow.  When I do, I'll let you know what happens.

**Reply by dresco** [reply]
> I am using Trinamic stepper motor drivers, three TMC2209s. The macros TRINAMIC_ENABLE 2209 and TRINAMIC_R_SENSE 110 are uncommented in my_machine.h. The HAS_BOARD_INIT macro was enabled when I compiled the previous version of grblHAL (pre-homing pulloff plugin) and there were no issues with compiling.

You'll want a board_init() function somewhere to initialise the Trinamic drivers. For the Morpho map, it's all in [this](https://github.com/grblHAL/STM32F4xx/blob/master/boards/st_morpho.c) file, but if your board definition has changed it won't be used.

Other boards use a simpler version for tmc uart init only;
https://github.com/grblHAL/STM32F4xx/blob/ffa25639089726f3c60937a397c3eeac7e51a91b/Src/tmc_uart.c#L485-L492

**Reply by terjeio** [reply]
I'll move all Trinamic SPI interfaces to _tmc_spi.c_ and call the init function from _driver.c_ in the next commit. HAS_BOARD_INIT will be removed from the boards that no longer need it.

**Reply by hollingerc** [reply]
I had a look at the board board_init() function, I'm using TMC2209 drivers and they use the UART interface, so the SPI interface is not needed.  I see the board_init() function will initialize the SPI or UART depending on the configuration.

I think my problems may have to do with me recently implementing a My Machine build.  I didn't implement any initialization specific to my board.  Don't really want to do that.

Previously, I was using the ST Morpho build with modified pin configuration in st_morpho_map.h.  This has worked for me, so I'm in the process of changing everything back.  I've had a successful build (no errors) with this configuration and the new code you just released.  I'll get back to you when I've had a chance to test it.

**Reply by hollingerc** [reply]
Have now done some preliminary testing of grbl with the new plugin and the $29x settings seem to work as expected.  Thanks for creating the new plugin.

But I still have a problem with homing the Z-axis due to the homing speeds.    The Z-axis needs much higher speeds than those for the X and Y axes.  I have set the $18x and $19x settings to the Z-axis speeds I need, but the actuator is moving much too slowly.  grbl seems to be applying only the $24 and $25 settings to determine the homing speeds, not what I put into the $18x and $19x settings.  The Trinamic plugin is working with the TMC2209 stepper motor drivers I have on my board, the motors spin and the actuators make the correct moves.  Is there something I've missed with using the $18x and $19x settings?  Do they need to be enabled somehow?

**Reply by terjeio** [reply]
>  Do they need to be enabled somehow?

They are only applied in conjunction with sensorless homing.

**Reply by hollingerc** [reply]
Oh, that's too bad.  I need different homing speeds for the Z-axis.  I thought that was what these settings were for.  Any chance of another plugin?

BTW, where are these settings documented?  The information I have for them says nothing about sensorless homing.

**Reply by hollingerc** [reply]
It would be useful to be able to set different homing speeds (for both locate and search seek) for each axis during normal limit switch homing, much like $18x and $19x do for sensorless homing.  Would you be interested in creating another plugin for different homing speeds for each axis as you recently did for the homing pulloff ($29x) setting?

**Reply by terjeio** [reply]
I have just [committed a new version](https://github.com/grblHAL/core/blob/master/changelog.md#20241208) with support for per axis homing feedrates. Enable with a flag in the $22 setting:

![image](https://github.com/user-attachments/assets/a733e59f-d5f3-404d-9192-2047cb9a8256)

The Web Builder is not yet updated to this version.

**Reply by hollingerc** [reply]
Great!  Thank you.  I'll check out the new version and let you know how it works out.

**Reply by hollingerc** [reply]
I installed, complied and FLASHed the new version.  I updated the board map and my_machine.h to what I was using previously, but it appears the Trinamic driver is not working.  This is the output from grbl after power up/reset:

GrblHAL 1.1f ['$' or '$HELP' for help]
[MSG:Warning: Could not communicate with stepper driver!]

When I tried to change $140, I got this error:

$140=1000
error:2

When I try to read the $18x and $19x settings, I get this error:

$180
error:53

This is the output from the $I command:

$I
[VER:1.1f.20241208:]
[OPT:VNMSL,100,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,ES,SED,PID]
[FIRMWARE:grblHAL]
[SIGNALS:HSEP]
[NVS STORAGE:*FLASH]
[FREE MEMORY:91K]
[DRIVER:STM32F411]
[DRIVER VERSION:241208]
[BOARD:ITL9700 Controller]
[AUX IO:6,5,0,0]
[PLUGIN:Trinamic v0.21]
[PLUGIN:Homing pulloff v0.01]
ok

Has something changed in the Trinamic driver?  I noticed a new revision number.  Has the serial port configuration changed?  I didn't change the configuration in the board map file. This is the configuration from my board map file I've been using for the last functional version of grbl:

#define SERIAL_PORT     2   // GPIOA: TX = 2, RX = 3
#define SERIAL1_PORT    1   // GPIOA: TX = 9, RX = 10

I don't think I've missed or changed anything.  The last time I updated grbl, everything went smoothly.  I don't think it's a hardware problem, I just FLASHed my previous version of grbl and didn't get the communications warning.

What's changed?

**Reply by terjeio** [reply]
> When I try to read the $18x and $19x settings, I get this error:
$180
error:53

These settings are only available when enabled in the $22 setting (bit 9):

```
$$=22
$22: Homing cycle as bitfield where setting bit 0 enables the rest:
    0 - Enable (1)
    1 - Enable single axis commands (2)
    2 - Homing on startup required (4)
    3 - Set machine origin to 0 (8)
    4 - Two switches shares one input (16)
    5 - Allow manual (32)
    6 - Override locks (64)
    7 - Keep homed status on reset (128)
    8 - Use limit switches (256)
    9 - Per axis feedrates (512)
```

> Has something changed in the Trinamic driver?

Yes, but not driver communication since that is not a part of it.

> I noticed a new revision number. Has the serial port configuration changed?

Yes, the default port assignment algorithm has been changed so your issue could be related to that.
To fix this you should explicitly allocate the UART port to the Trinamic drivers by adding
`#define TRINAMIC_STREAM 0` or
`#define TRINAMIC_STREAM 1`
to your map file. You can check the assignment with the `$pins` command.

**Reply by hollingerc** [reply]
Adding the TRINAMIC_STREAM macro to my map file didn't eliminate the communications warning.  Before I added the macro, I checked the assignment with $pins:

[PIN:PA3,RX,Primary UART]
[PIN:PA2,TX,Primary UART]
[PIN:PA10,RX,Trinamic UART]
[PIN:PA9,TX,Trinaic UART]

The PA9 and PA10 pins are what I have the Trinamic drivers connected to.

I added #define TRINAMIC_STREAM 1 to my map file, re-compiled and FLASHed and still got the communications warning.  The pin assignments were the same as well.

When I changed the maco to 0, I still got the same communications warning and the pin assignment changed to:

[PIN:PA3,RX,Primay UART]
[PIN:PA2,TX,Primary UART]
[PIN:PA10,RX,UART2]
[PIN:PA9,TX,UART2]

This appears to not be the correct assignment.

I did a quick search for TRINAMIC_STREAM through the grbl firmware and found that the macro is defined to 1 if it is previously undefined.  It seems that I don't need to explicitly define it in my map file.

Anyway, this macro didn't eliminate the communications warning.  Is there some other macro I need to add or some $ setting I should change?

**Reply by terjeio** [reply]
> I updated the board map and my_machine.h to what I was using previously

If you are modifying an existing board map then it is better to copy it to _my_machine_map.h_ instead, modify it and enable that in _my_machine.h_ instead, this way it will not be overwritten on an update.

When I test with my BOARD_MORPHO_CNC the TMC2209 drivers works as they should, connected to PA9 and P10.
Since the previous version works I assume the drivers are powered (motor power) at the same time as/before the controller. Another possible reason for failure that comes to mind is that PA9 or PA10 has been accidentally assigned to a different function. This has happened to me before.

**Reply by hollingerc** [reply]
I did try using my_machine_map.h, but I had a problem board initialization.  We discussed this issue in another post.  I didn't want to write my own board initialization function.  I'm not too bothered about updating the st_morpho_map.h file, I don't expect to be updating grbl in the future once this Z-axis homing issue gets resolved.

I don't know how PA9 and PA10 could get accidentally re-assigned, I didn't make any changes to the st_morpho_map.h file other than adding the TRINAMIC_STREAM macro.  I checked the pin assignments with $pins and these two pins are assigned only to the Trinamic UART.

BTW, this is how I update the grblHAL firmware:
 - git clone the grblHAL repository with "git clone --recurse-submodules https://github.com/grblHAL/STM32F4xx.git"
 - rename the new folder to my project name
 - import the project into STM32CubeIDE
 - make the changes to my_machine.h and st_morpho_map.h 
 - compile and FLASH

This seems to me to be a foolproof way of updating the firmware.  The newest version I'm working with, the one that gives the communications warning, is [VER:1.1f.20241208:].

I keep the older versions of the grblHAL that I was working with, I just rename the project folders by adding the date that I update to a new version.  This way I've been able to quickly go back to older versions.

If this version worked for you, why isn't it working for me?  Any other ideas?

**Reply by terjeio** [reply]
> I did try using my_machine_map.h, but I had a problem board initialization.

Trinamic driver initialization was recently moved to _driver.c_, `#define HAS_BOARD_INIT` is no longer required in the board map unless there is other custom stuff that needs to be done. So the board init problem should now be resolved.

> The newest version I'm working with, the one that gives the communications warning, is [VER:1.1f.20241208:].

The previous that works is which version? And this is with the same controller board that fails with the latest version?

> If this version worked for you, why isn't it working for me? Any other ideas?

The warning is issued if one of the drivers is unresponsive. If you only enable one driver, e.g. X, with $338=1 does it still fail?

If you flash the debug version can you try with setting breakpoints in [TMC2209_Init()](https://github.com/terjeio/Trinamic-library/blob/47f58689a01b9676590e736b8c0ef7e6b1c6ed9c/tmc2209.c#L171-L206):
Is the function entered at all?
Does it reach the first read register call in line 178.
Differs the `reg.count` and `ifcount` values by 7 in line 205?

**Reply by hollingerc** [reply]
OK, I'll move the contents of st_morpho_map.h to my_machine_map.h in the future.

The previous working version I was using is [VER:1.1f.20241113:].  I believe this was the version that implemented separate homing pull off distances for each axis.

I'll try the debugging suggestions shortly and get back to you.

**Reply by hollingerc** [reply]
Each time I change $338, the same message comes up (as before):

[MSG:Warning: Could not communicate with stepper driver!]

When running the debugger, line 178 in TMC2209.c is not reached after reset.  The return on line 176 is executed.

If I hang a scope on the USART TX pin of the STM32F411 all I see after reset is a single negative 5msec long pulse.  The line idles high.  Seems to me the UART in the F411 is not initialized properly as there is no data being transmitted.  I'd guess the tx pin was configured correctly since it is idling in the high state.

Could there be an issue with my configuration causing the UART to not get initialized?

**Reply by terjeio** [reply]
> Seems to me the UART in the F411 is not initialized properly as there is no data being transmitted. I'd guess the tx pin was configured correctly since it is idling in the high state.

I have just checked with a F411 since I used a F446 earlier, both it works for me.
The UART is configured and claimed early in the startup sequence, before the rest of the GPIOs. If a UART pin is later reconfigured for a different function then the UART itself works but the output (or input) never reaches its destination. As I said earlier I have had this happening to me during development. So either I have introduced a bug somewhere or you have a typo in your board map.

BTW only when the port has been initialized and claimed will the $pins command show that the pins are assigned as `Trinamic UART. That is why I believe the issue is not related to the UART code itself.

Can you post your board map file?

**Reply by hollingerc** [reply]
Here is my board map file.  It has worked through the last two revisions of grbl and the only change I made (other than the comments) for the latest version is to add the TRINAMIC_STREAM macro.

/*
  st_morpho_map.h - driver code for STM32F411 on ITL9700 controller board

  Part of grblHAL

  Copyright (c) 2020-2024 Terje Io

  grblHAL is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  grblHAL is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with grblHAL. If not, see <http://www.gnu.org/licenses/>.
*/
/*--------------------------------------------------------------------------------------------------
 *
 * Modified by: Craig Hollinger
 *              for the ITL9700 refresh project
 *              Intralock Tools Limited
 *
 *--------------------------------------------------------------------------------------------------
 * Date         Description
 *--------------------------------------------------------------------------------------------------
 * 30-Oct-2024  Enabled BOARD_MY_MACHINE in my_machine.h, added my_machine_map.h
 *
 * 19-Nov-2024  New version of grblHAL (1.1f.20241113) which put the homing pull off distance into
 *              separate settings ($29x).  For this version to work, a new macro,
 *              HOMING_PULLOFF_ENABLE, was defined in my_machine.h.  This version didn't work with a
 *              My Machine build, needs a board_init() function.  Changed back to ST Morpho build.
 *
 * 10-Dec-2024  New version of grblHAL (1.1f.20241208) which puts the two homing feed rates into
 *              separate settings ($18x and $19x) for each axis.  This version brought up an error
 *              indicating problems with communicating with the Trinamic controllers.  Changes to
 *              the Trinamic driver required a new macro TRINAMIC_STREAM to be defined in this file.
 *------------------------------------------------------------------------------------------------*/

#if TRINAMIC_ENABLE == 2130
#include "trinamic/tmc2130.h"
#endif

#if TRINAMIC_ENABLE == 2209
#include "trinamic/tmc2209.h"
#endif

#if N_ABC_MOTORS > 1
#error "Axis configuration is not supported!"
#endif

#if SDCARD_ENABLE && ETHERNET_ENABLE
#error "SD card and networking plugins cannot be enabled together!"
#endif

#define BOARD_NAME "ITL9700 Controller"
#define BOARD_URL ""

/*
#ifndef WEB_BUILD
#undef I2C_ENABLE
#undef EEPROM_ENABLE
//#undef EEPROM_IS_FRAM
#define I2C_ENABLE      1
#define EEPROM_ENABLE   16
//#define EEPROM_IS_FRAM  1
#endif
*/
#define SERIAL_PORT     2   // GPIOA: TX = 2, RX = 3
#define SERIAL1_PORT    1   // GPIOA: TX = 9, RX = 10
#if I2C_ENABLE
#define I2C_PORT        1   // GPIOB: SCL = 8, SDA = 9
#endif

/* This new macro allocates the UART to the Trinamic drivers.  Could be 0 or 1. */
#define TRINAMIC_STREAM 1

#define IS_NUCLEO_BOB
#define HAS_BOARD_INIT

#if SDCARD_ENABLE || TRINAMIC_SPI_ENABLE || ETHERNET_ENABLE
#define SPI_PORT                1 // GPIOA, SCK_PIN = 5, MISO_PIN = 6, MOSI_PIN = 7
#endif

// Define step pulse output pins.
#define STEP_PORT               GPIOB
#define X_STEP_PIN              8
#define Y_STEP_PIN              6
#define Z_STEP_PIN              4
#if N_GANGED
#define STEP_OUTMODE            GPIO_MAP
#else
#define STEP_OUTMODE            GPIO_MAP
#endif
//#define STEP_PINMODE            PINMODE_OD // Uncomment for open drain outputs

// Define step direction output pins.
#define DIRECTION_PORT          GPIOB
#define X_DIRECTION_PIN         7
#define Y_DIRECTION_PIN         5
#define Z_DIRECTION_PIN         3
#if N_GANGED
#define DIRECTION_OUTMODE       GPIO_BITBAND
#else
#define DIRECTION_OUTMODE       GPIO_MAP
#endif
//#define DIRECTION_PINMODE       PINMODE_OD // Uncomment for open drain outputs

// Define stepper driver enable/disable output pins.

#define X_ENABLE_PORT           GPIOC
#define X_ENABLE_PIN            12
#define Y_ENABLE_PORT           GPIOC
#define Y_ENABLE_PIN            11
#define Z_ENABLE_PORT           GPIOC
#define Z_ENABLE_PIN            10
//#define STEPPERS_ENABLE_PINMODE PINMODE_OD // Uncomment for open drain outputs

// Define homing/hard limit switch input pins.
#define LIMIT_PORT              GPIOA
#define X_LIMIT_PIN             11
#define Y_LIMIT_PIN             15
#define Z_LIMIT_PIN             12
#define LIMIT_INMODE            GPIO_BITBAND

// Define ganged axis or A axis step pulse and step direction output pins.
#if N_ABC_MOTORS == 1
#define M3_AVAILABLE
#define M3_STEP_PORT            GPIOC
#define M3_STEP_PIN             6
#define M3_DIRECTION_PORT       GPIOA
#define M3_DIRECTION_PIN        12
#define M3_LIMIT_PORT           GPIOC
#define M3_LIMIT_PIN            11
#define M3_ENABLE_PORT          GPIOB
#define M3_ENABLE_PIN           6
#endif

// Define driver spindle pins

#if DRIVER_SPINDLE_PWM_ENABLE
#define SPINDLE_PWM_PORT_BASE   GPIOA_BASE
#define SPINDLE_PWM_PORT        AUXOUTPUT0_PORT
#define SPINDLE_PWM_PIN         8
#else
#define AUXOUTPUT4_PORT         GPIOA
#define AUXOUTPUT4_PIN          8
#endif

#if DRIVER_SPINDLE_DIR_ENABLE
#define SPINDLE_DIRECTION_PORT  GPIOA
#define SPINDLE_DIRECTION_PIN   1
#else
#define AUXOUTPUT5_PORT         GPIOA
#define AUXOUTPUT5_PIN          7
#endif

#if DRIVER_SPINDLE_ENABLE
#define SPINDLE_ENABLE_PORT     GPIOA
#define SPINDLE_ENABLE_PIN      0
#else
#define AUXOUTPUT6_PORT         GPIOA
#define AUXOUTPUT6_PIN          0
#endif

// Define flood and mist coolant enable output pins.
#define COOLANT_FLOOD_PORT      GPIOC
#define COOLANT_FLOOD_PIN       2
#define COOLANT_MIST_PORT       GPIOC
#define COOLANT_MIST_PIN        3

// Define user-control controls (cycle start, reset, feed hold) input pins.
#define CONTROL_PORT            GPIOC
#define RESET_PIN               4
#define FEED_HOLD_PIN           5
#define CYCLE_START_PIN         6
#define CONTROL_INMODE          GPIO_BITBAND

// Auxiliary I/O

#define AUXOUTPUT0_PORT         GPIOC
#define AUXOUTPUT0_PIN          0
#if !ETHERNET_ENABLE
#define AUXOUTPUT1_PORT         GPIOC
#define AUXOUTPUT1_PIN          1
#endif

#undef I2C_STROBE_ASSIGNED

#ifdef SPI_PORT

#if SDCARD_ENABLE
#define SD_CS_PORT              GPIOA
#define SD_CS_PIN               4
#endif

#if TRINAMIC_SPI_ENABLE
#define MOTOR_CS_PORT           GPIOC
#define MOTOR_CS_PIN            7
#endif

#if ETHERNET_ENABLE
#undef SPI_ENABLE
#define SPI_ENABLE 1
#define SPI_CS_PORT             GPIOC
#define SPI_CS_PIN              13
#define SPI_IRQ_PORT            GPIOC
#define SPI_IRQ_PIN             14
#define SPI_RST_PORT            GPIOC // AUXOUTPUT1
#define SPI_RST_PIN             15
#endif

#else

#define AUXOUTPUT2_PORT         GPIOA
#define AUXOUTPUT2_PIN          5
#define AUXOUTPUT3_PORT         GPIOA
#define AUXOUTPUT3_PIN          6
#ifndef AUXOUTPUT4_PORT
//#define AUXOUTPUT4_PORT         GPIOA
//#define AUXOUTPUT4_PIN          7
#endif
#ifndef AUXOUTPUT5_PORT
#define AUXOUTPUT5_PORT         GPIOA
#define AUXOUTPUT5_PIN          7
#endif
#if MPG_MODE == 0
#if 0
#define AUXOUTPUT0_PWM_PORT_BASE GPIOC_BASE
#define AUXOUTPUT0_PWM_PIN       8
#elif !defined(AUXOUTPUT6_PORT)
#define AUXOUTPUT6_PORT         GPIOC
#define AUXOUTPUT6_PIN          8
#endif
#endif
#endif

/*
#define LED_R_PORT              GPIOA
#define LED_R_PIN               7
#define LED_G_PORT              GPIOB
#define LED_G_PIN               7
#define LED_B_PORT              GPIOC
#define LED_B_PIN               8
*/

//#define NEOPIXEL_SPI            1 // PA7
//#define NEOPIXELS_NUM           6

#if !SPINDLE_SYNC_ENABLE && !QEI_ENABLE
#define AUXINPUT0_PORT          GPIOB
#define AUXINPUT0_PIN           0
#endif

#if !SPINDLE_ENCODER_ENABLE && !QEI_ENABLE
#define AUXINPUT1_PORT          GPIOB
#define AUXINPUT1_PIN           1
#endif

#ifndef M3_LIMIT_PIN
#define AUXINPUT2_PORT          GPIOB
#define AUXINPUT2_PIN           2
#endif

#define AUXINPUT3_PORT          GPIOB
#define AUXINPUT3_PIN           9
#define AUXINPUT4_PORT          GPIOB
#define AUXINPUT4_PIN           10
#define AUXINPUT5_PORT          GPIOB
#define AUXINPUT5_PIN           12

#ifndef SPI_PORT
#define AUXINPUT6_PORT          GPIOB
#define AUXINPUT6_PIN           13
#elif N_ABC_MOTORS == 0
#define AUXINPUT6_PORT          GPIOB
#define AUXINPUT6_PIN           13
#endif

#if PROBE_ENABLE
#define PROBE_PORT              AUXINPUT5_PORT
#define PROBE_PIN               AUXINPUT5_PIN
#endif

#if SAFETY_DOOR_ENABLE
#define SAFETY_DOOR_PORT        AUXINPUT4_PORT
#define SAFETY_DOOR_PIN         AUXINPUT4_PIN
#endif

#if MOTOR_FAULT_ENABLE && defined(AUXINPUT1_PORT)
#define MOTOR_FAULT_PORT        AUXINPUT1_PORT
#define MOTOR_FAULT_PIN         AUXINPUT1_PIN
#endif

#if MOTOR_WARNING_ENABLE && !I2C_STROBE_ENABLE
#define MOTOR_WARNING_PORT      AUXINPUT3_PORT
#define MOTOR_WARNING_PIN       AUXINPUT3_PIN
#endif

#if I2C_STROBE_ENABLE
#define I2C_STROBE_PORT         AUXINPUT3_PORT
#define I2C_STROBE_PIN          AUXINPUT3_PIN
#endif

#if MPG_MODE == 1
#define MPG_AUX_ENABLE
#define MPG_MODE_PORT           AUXINPUT6_PORT
#define MPG_MODE_PIN            AUXINPUT6_PIN
#endif

#if SPINDLE_ENCODER_ENABLE
#define SPINDLE_PULSE_PORT      GPIOB  // NOTE: this pin is connected to PA15
#define SPINDLE_PULSE_PIN       14
#if SPINDLE_SYNC_ENABLE
#define SPINDLE_INDEX_PORT      GPIOB
#define SPINDLE_INDEX_PIN       15
#endif
#endif

#if QEI_ENABLE && !SPINDLE_SYNC_ENABLE
#define QEI_A_PORT              GPIOC
#define QEI_A_PIN               8
#define QEI_B_PORT              GPIOC
#define QEI_B_PIN               9
#endif // QEI_ENABLE

#if QEI_SELECT_ENABLE
#if !I2C_STROBE_ENABLE
#define QEI_SELECT_PORT         AUXINPUT3_PORT
#define QEI_SELECT_PIN          AUXINPUT3_PIN
#define I2C_STROBE_ASSIGNED
#elif !SAFETY_DOOR_ENABLE
#define QEI_SELECT_PORT         AUXINPUT4_PORT
#define QEI_SELECT_PIN          AUXINPUT4_PIN
#endif
#endif

// EOF

**Reply by terjeio** [reply]
Your map file above gave compilation errors, when fixed I got the same communication error. After commenting out general GPIO inits it still failed so I decided to try a settings reset with `$RST=*` and that did the trick, at least with the one driver (X) I have wired up on a breadboard. I did not save the original flash content so I have no idea about what in the original settings that caused the error.
While doing this I also power cycled the driver (motor power) a few times to ensure it was in a fresh state.

As a final test I erased the flash with the ST-Link utility, after reflashing the firmware worked without resetting the settings.

Here is the modified board map, renamed to _my_machine_map.h_, that works:
[my_machine_map.zip](https://github.com/user-attachments/files/18099558/my_machine_map.zip)

**Reply by hollingerc** [reply]
Thanks for working on this.  

I'll incorporate the new my_machine_map.h file and have a go.  Let you know what happens.

**Reply by hollingerc** [reply]
I did a file compare between the my_machine_map.h you sent me and my st_morpho_map.h.  The difference I could see is you took out the first two #include statements and the HAS_BOARD_INIT macro.  The HAS_BOARD_INIT macro was left over from when I tried using my own my_machine_map.h file, I didn't intend to leave it there.  I made these changes, complied and FLASHed the new firmware and the communications warning is no longer sent.  The $ settings are all at their defaults.  If I send a M122 command the correct Trinamic driver information is sent.  Looks like the communications to the Trinamic drivers is working.

I updated all the settings I was using, including the $18 and $19x, to the values I use and the Z-axis homing worked as I wanted (as did the other two axes).  Looks like the new version of grbl is working correctly.  It would seem the errant HAS_BOARD_INIT macro was the culprit with the communications warning.

Thanks a lot for modifying grbl as per my requests.  You work is much appreciated.

BTW, do you accept donations?  If so, how can I send you some money?  You deserve some remuneration for your work on this project.

**Reply by hollingerc** [reply]
Were there any changes around the spindle driver with this new version of grbl (20241208)?  I was just doing some final testing and found that the spindle motor is not turned on when the M3 command is sent.  This was working with an older version of grbl (20240228).  The MCU GPIO pin does not change state.  This problem occurs with the previous version of grbl (20241113) as well.

Is there a new macro that has to be added somewhere?

Below are the spindle GPIO defines from my board map file:

// Define driver spindle pins

#if DRIVER_SPINDLE_PWM_ENABLE
#define SPINDLE_PWM_PORT_BASE   GPIOA_BASE
#define SPINDLE_PWM_PORT        AUXOUTPUT0_PORT
#define SPINDLE_PWM_PIN         8
#else
#define AUXOUTPUT4_PORT         GPIOA
#define AUXOUTPUT4_PIN          8
#endif

#if DRIVER_SPINDLE_DIR_ENABLE
#define SPINDLE_DIRECTION_PORT  GPIOA
#define SPINDLE_DIRECTION_PIN   1
#else
#define AUXOUTPUT5_PORT         GPIOA
#define AUXOUTPUT5_PIN          7
#endif

#if DRIVER_SPINDLE_ENABLE
#define SPINDLE_ENABLE_PORT     GPIOA
#define SPINDLE_ENABLE_PIN      0
#else
#define AUXOUTPUT6_PORT         GPIOA
#define AUXOUTPUT6_PIN          0
#endif

**Reply by dresco** [reply]
> Were there any changes around the spindle driver with this new version of grbl (20241208)? I

Yes, the spindle outputs moved to claimed aux ports earlier this year. If you look at the current [morpho map file](https://github.com/grblHAL/STM32F4xx/blob/2f13bf43b4c22147103f2decebd731fd7aae318e/boards/st_morpho_map.h#L128-L151), should give you an idea of the changes needed to yours..

**Reply by hollingerc** [reply, marked as answer]
I modified the st_morpho_map.h file you sent me and cloned the latest version of grblHAL.  It compiled without errors.  The spindle output now responds to M3 and M5 commands.

Thanks again.
